R/bundle.R

```diff
@@ -517,6 +517,14 @@ inferRPackageDependencies <- function(appMode, hasParameters, documentsHavePytho
   unique(deps)
 }
 
+# shiny::renderPlot() now uses ragg for png rendering if it's installed,
+# so include it if it's installed locally (this way users will get consistent
+# PNG results without needing to add library(ragg) to the app).
+# https://github.com/rstudio/shiny/pull/3654
+shinyDeps <- function() {
+  c("shiny", if (system.file(package = "ragg") != "") "ragg")
```

Comment:
Is there any merit to narrowing this further? For example, add `ragg` only when the Shiny version is >= 1.7.1 (or whatever)?

Is there a plan to make ragg the default (and therefore an explicit dependency)?

Reply:
With this explanation, I'm OK without the version check.

Do you think there are cases where folks will want to opt-out of `ragg`? In other words: Are there systems that will not be able to install `ragg`? This change is adding it as an explicit requirement to the manifest / packrat lock -- which means that deploys will err if those systems cannot build / install `ragg`.

Reply:
Hm, yes, it is possible that systems won't be able to install ragg. [ragg](https://cran.r-project.org/web/packages/ragg/) imports [systemfonts](https://cran.r-project.org/web/packages/systemfonts/) and [textshaping](https://cran.r-project.org/web/packages/textshaping/). Here are the system requirements for each:

* ragg: C++11, freetype2, libpng, libtiff, libjpeg
* systemfonts: C++11, fontconfig, freetype2
* textshaping: C++11, freetype2, harfbuzz, fribidi

Up until this point, I thought that we could assume that the deployment targets could build/install these packages. Is that not the case?

IIRC, we used to install Cairo by default on shinyapps.io because of plot rendering issues. Do you remember if that's the case? It would be nice if we could do something similar today with ragg (but of course it wouldn't just be on shinyapps.io).

Reply:
If we're worried about some environments not being able to install `{ragg}` (which seems like a valid concern since IIRC these packages have had issue in the past with older R versions), maybe it is worth adding the shiny version check.

Reply:
Just to echo the concern of `ragg` having some not-so-standard system dependencies (`textshaping` requires harfbuzz and fribidi), I just tried to re-deploy an app to a connect server with nontrivial updates to the app (some changes in text) and the deploy failed (did not fail a few days prior) because harfbuzz and fribidi weren't available on the server.

It took a lot of head scratching and digging to figure out that the only thing that had changed between app deployments is that `ragg` was not installed locally before the failed deployment and at some point got installed (not because of an app dependency) before the failed deployment. Because `rsconnect` found `ragg` locally it then added this as a dependency, which resulted in the failed deployment. I had a hard time tracking down why `ragg` and `textshaping` were in my packrat snapshot.

Anyway, that's a longwinded way to say that it might be good to at least provide a message when extra dependencies are added that are not inferred directly from the app code to help people understand why these dependencies are there.

Reply:
Echoing hafen. We had exactly the same issue with a mysterious failed deployment. We control our dev environment very carefully with an renv lockfile, and we could not understand why rsconnect had randomly decided that we needed to install textshaping when we hadn't asked it to do so. I spent an hour playing detective in the log files and the google search bar until I came to the rsconnect NEWS.md item about the new implicit behavior in rsconnect 0.8.27.

Reply:
The addition of `ragg` has been reverted with https://github.com/rstudio/rsconnect/pull/625, which should be released to CRAN soon.

Reply:
Thank you - I am fighting for RSConnect/Shiny to be considered a legitimate production environment at my company, and I think this attention to engineering quality will reassure everyone that we won't regret the decision.